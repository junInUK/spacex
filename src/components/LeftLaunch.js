import React, { Component } from "react";
import launch_home from '../assets/img/launch-home.png';
import '../App.css';

class LeftLaunch extends Component{
    render(){
        return (
            <div className="row left-launch">
                <img src={launch_home} width="100%" alt="launch home"></img>
            </div>
        )
    }
}

export default LeftLaunch;