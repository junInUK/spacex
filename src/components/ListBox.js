import React, { Component } from "react";
import LaunchDetail from './LaunchDetail';
import '../App.css';

class ListBox extends Component{
    constructor(props){
        super(props);

        this.state = {
            matchedYearLaunchLists: []
        }

        this.selectYear = this.selectYear.bind(this);
        this.descendSort = this.descendSort.bind(this);
    }

    componentDidMount() {
        this.setState({ matchedYearLaunchLists: this.props.launchLists });
        console.log("matchedLaunchList length:"+ this.state.matchedYearLaunchLists.length);
    }

    descendSort(){
        console.log("Yes, descend launch list!");
        this.props.descendSort();
    }

    selectYear(event){
        console.log("Yes, select year?"+event.target.value);
        this.props.selectYear(event.target.value);
    }

    render(){
        const noRepeatLaunchYearsWithA = this.props.noRepeatLaunchYears.map(launchYear => {
            return (
                // <a href="#" onClick={this.selectYear}>{launchYear}</a>
                <option className="yearOption" value={launchYear} key={launchYear}>{launchYear}</option>
            );
        });
        
        const launchLists = this.props.launchLists.map(launchDetail => {
            return (
                <LaunchDetail flightNumber = {launchDetail.flight_number}
                missionName = {launchDetail.mission_name}
                launchDateUTC = {launchDetail.launch_date_utc}
                rocketName = {launchDetail.rocket.rocket_name}
                key = {launchDetail._id}
                />
            );
        });

        return (
            <div className="container list-box">
                <div className="row list-button">
                    <div className="dropdown">
                        <button className="btn btn-info fliter-year">
                            Filter by Year <span className="icon-select"> </span>
                        </button>
                        <select className="dropdown-year" onChange={this.selectYear}>
                            {noRepeatLaunchYearsWithA}
                        </select>
                        {/* <div className="dropdown-year"> */}
                            
                        {/* </div> */}
                        
                    </div>
                    <button className="btn btn-info sort" onClick={this.descendSort}>
                        Sort Descending <span className="icon-sort"> </span>
                    </button>
                </div>
                {launchLists}
            </div>
        )
    }
}

export default ListBox;