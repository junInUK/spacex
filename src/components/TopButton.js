import React, { Component } from "react";
import '../App.css';
import spacex_logo from '../assets/spacex-logo.png';

class TopButton extends Component{
    constructor(props){
        super(props);
        
        this.reLoadData = this.reLoadData.bind(this);
    }

    reLoadData(){
        this.props.pReloadData();
    }

    render(){
        return (
            <div className="row top-button">
                <div className="col-sm-9 col-md-9 col-lg-6">
                    <img src={spacex_logo} width="25%" alt="spacex logo"></img>
                    <span className="text-after-logo">LAUNCHES</span>
                </div>
                <div className="col-sm-9 col-md-9 col-lg-6 text-right">
                    <button className="btn btn-info" onClick={this.reLoadData}>Reload Data <span className="icon-refresh"> </span></button>
                </div>
            </div>
        )
    }
}

export default TopButton;