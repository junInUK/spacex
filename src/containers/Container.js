import React, { Component } from 'react';
import TopButton from '../components/TopButton';
import LeftLaunch from '../components/LeftLaunch';
import ListBox from '../components/ListBox';

class Container extends Component {
    constructor(props){
        super(props);

        this.state = {
            launchLists: [],
            noRepeatLaunchYears: [],
            filteredYear: 0
        }
        this.reloadData = this.reloadData.bind(this);
        this.selectYear = this.selectYear.bind(this);
        this.descendSort = this.descendSort.bind(this);
    }

    loadData(){
        fetch("https://api.spacexdata.com/v3/launches?id=true")
        .then(res => res.json())
        .then(result => {
            this.setState({launchLists: result});
            let years = [];
            for(let i = 0; i < this.state.launchLists.length; i++){
                console.log(years.indexOf(this.state.launchLists[i].launch_date_utc.substring(0,4)));
                if(years.indexOf(this.state.launchLists[i].launch_date_utc.substring(0,4)) === -1){
                    years.push(this.state.launchLists[i].launch_date_utc.substring(0,4));
                }
            }
            this.setState({noRepeatLaunchYears: years});
        })
        .catch(err => console.log(err)) 
    }

    selectYear(yearSelected){
        console.log("In container, year selected:"+yearSelected);
        let matchedYearLaunchLists = [];
        for(let i = 0; i< this.state.launchLists.length; i++){
            let year = this.state.launchLists[i].launch_date_utc.substring(0,4);
            console.log("launched year:["+ year+"]");
            if(year === yearSelected){
                console.log(year + " == " + yearSelected)
                matchedYearLaunchLists.push(this.state.launchLists[i]);
            }
        }
        console.log("after match:");
        for(let i = 0; i < matchedYearLaunchLists.length; i++){
            console.log(matchedYearLaunchLists[i]);
        }
       
        this.setState({
            launchLists: matchedYearLaunchLists
        });
        
        console.log("In state launch list:");
        for(let i = 0; i<this.state.launchLists.length; i++){
            console.log(this.state.launchLists[i].launch_date_utc);
        }
    }

    descendSort(){
        console.log("In container, descend sort!");
        let tempLaunchLists = [...this.state.launchLists];
        tempLaunchLists.sort(function(a,b){
            let aYear = a.launch_date_utc.substring(0,4);
            let aMonth = a.launch_date_utc.substring(5,7);
            let aDate = a.launch_date_utc.substring(8,10);
            let bYear = b.launch_date_utc.substring(0,4);
            let bMonth = b.launch_date_utc.substring(5,7);
            let bDate = b.launch_date_utc.substring(8,10);
            console.log("a year:"+aYear+ " a month:" + aMonth + " a date:" + aDate +
                       " b year:"+bYear+ " b month:" + bMonth + " b date:" + bDate);
            if(aYear < bYear){
                return 1;
            }else if(aYear > bYear){
                return -1;
            }else{
                if(aMonth < bMonth){
                    return 1;
                }else if(aMonth > bMonth){
                    return -1;
                }else{
                    if(aDate < bDate){
                        return 1;
                    }else if(aDate > bDate){
                        return -1;
                    }else{
                        return 0;
                    }
                }
            }    
        });
        for(let i=0;i<tempLaunchLists.length;i++){
            console.log(tempLaunchLists[i].launch_date_utc);
        }
        this.setState({
            launchLists: tempLaunchLists
        })
    }

    reloadData(){
        console.log("This is reloadData in the container!");
        this.loadData()
    }

    componentDidMount() {
        this.loadData()
    }

    render() {
        return (
            <div className="container">
                <TopButton pReloadData={this.reloadData}/>      
                <div className="row">
                    <div className="col-sm-9 col-md-9 col-lg-5">
                        <LeftLaunch />
                    </div>
                    <div className="col-sm-9 col-md-9 col-lg-7 text-right">
                        <ListBox launchLists = {this.state.launchLists}
                        noRepeatLaunchYears = {this.state.noRepeatLaunchYears} 
                        selectYear = {this.selectYear} 
                        descendSort = {this.descendSort} />
                    </div>
                </div>
            </div>
        )
    }
}

export default Container;