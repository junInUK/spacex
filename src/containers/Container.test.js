import React from 'react';
import Container from './Container';
import TopButton from '../components/TopButton';
import LeftLaunch from '../components/LeftLaunch';
import ListBox from '../components/ListBox';
import { shallow, mount } from 'enzyme';			

describe('Container', () => {
    let wrapper;
    beforeEach(() => wrapper = shallow(<Container />));

    it("should render four <div>", () => {
        expect(wrapper.find("div").length).toEqual(4);
    });

    it("should render a TopButton", () => {
        expect(wrapper.containsMatchingElement(<TopButton />)).toEqual(true);
    });

    it("should render a LeftLaunch", () => {
        expect(wrapper.containsMatchingElement(<LeftLaunch />)).toEqual(true);
    });

    it("should render a ListBox", () => {
        expect(wrapper.containsMatchingElement(<ListBox />)).toEqual(true);
    });

    it("should render TopButton, LeftLaunch and ListBox together", () => {
        expect(wrapper.containsAllMatchingElements([
            <TopButton />,
            <LeftLaunch />,
            <ListBox />
        ])).toEqual(true);
    });

    it("renders corrently", () => {
        expect(wrapper).toMatchSnapshot();
    });
});

describe("Mounted Container", () => {
    it("should call reloadData when the reload button is clicked", () => {
        const spy = jest.spyOn(Container.prototype, "reloadData");
        const wrapper = mount(<Container/>);
        const topButton = wrapper.find("TopButton");
        wrapper.find("#reload-button").simulate("click");
        expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should call selectYear when the Filter by Year button is clicked", () => {
        const spy = jest.spyOn(Container.prototype, "selectYear");
        const wrapper = mount(<Container/>);
        const ListBox = wrapper.find("ListBox");
        wrapper.find("#select-year").simulate("change");
        expect(spy).toHaveBeenCalledTimes(1);
    });

    it("should call descendSort when the Sort Descending button is clicked", () => {
        const spy = jest.spyOn(Container.prototype, "descendSort");
        const wrapper = mount(<Container/>);
        const ListBox = wrapper.find("ListBox");
        wrapper.find("#sort-descend").simulate("click");
        expect(spy).toHaveBeenCalledTimes(1);
    });

});