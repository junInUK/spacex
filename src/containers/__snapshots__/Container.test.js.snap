// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container renders corrently 1`] = `
<div
  className="container"
>
  <TopButton
    pReloadData={[Function]}
  />
  <div
    className="row"
  >
    <div
      className="col-sm-9 col-md-9 col-lg-5"
    >
      <LeftLaunch />
    </div>
    <div
      className="col-sm-9 col-md-9 col-lg-7 text-right"
    >
      <ListBox
        descendSort={[Function]}
        launchLists={Array []}
        noRepeatLaunchYears={Array []}
        selectYear={[Function]}
      />
    </div>
  </div>
</div>
`;
